File: scripts/python/project_health_mvg_versions.js
/* Read-only main snapshot navigation; render all repository text with textContent. */
(() => {
  const get = id => document.getElementById(id);
  const item = (tag, content, parent) => { const node = document.createElement(tag); node.textContent = String(content ?? ''); parent.append(node); return node; };
  const source = (path, parent) => { const link = item('a', path, parent); link.href = '/api/knowledge/source?path=' + encodeURIComponent(path); link.target = '_blank'; link.rel = 'noopener'; return link; };
  const list = (label, values, parent, links = false) => {
    const group = document.createElement('details'); group.open = false;
    item('summary', `${label} (${values.length})`, group);
    for (const value of values) { const line = document.createElement('p'); if (links && value.includes('/')) source(value, line); else item('span', value, line); group.append(line); }
    parent.append(group);
  };
  let overview;
  function render() {
    const version = overview.versions.find(row => row.id === get('mvg-version-select').value);
    const manifest = overview.manifests.find(row => row.path === get('mvg-manifest-select').value);
    const details = get('mvg-version-detail'); const flows = get('mvg-flow-detail'); details.replaceChildren(); flows.replaceChildren();
    if (version) {
      const heading = item('h3', version.gdd_path, details); heading.title = version.id;
      item('p', `GDD content SHA-256: ${version.gdd_sha256} · Chapter 3 trace: ${version.mapping} · main revision: ${overview.revision}`, details);
      source(version.gdd_path, details);
      if (version.mapping !== 'traced') item('p', 'Registered GDD candidate. Chapter 3 consumption and downstream ownership are not established in the published main topology.', details);
      list('Taskmaster primary IDs', version.tasks.map(task => `${task.id} · ${task.status || 'unknown'} · ${task.title || ''}`), details);
      for (const [key, label] of [['overlays', 'Overlay references'], ['contracts', 'Contract references'], ['adrs', 'ADR references']]) list(label, version.references[key] || [], details, key !== 'adrs');
      const scenes = document.createElement('details'); item('summary', `Statically attached scenes (${version.scenes.length})`, scenes);
      for (const scene of version.scenes) {
        const row = document.createElement('details'); item('summary', `${scene.path} · ${scene.classification || 'unknown'}`, row);
        item('p', 'Declared task association and verified static script attachment; this does not prove a runtime route.', row);
        const open = item('button', 'Open scene details', row); open.type = 'button';
        open.addEventListener('click', () => { if (window.openScenePreview) window.openScenePreview(scene.path); });
        list('Nodes', scene.nodes, row); list('Node resources', scene.resources, row, true); scenes.append(row);
      }
      details.append(scenes);
    }
    if (!manifest) { item('p', 'No MVG manifest in the scanned main snapshot.', flows); return; }
    const coverage = manifest.coverage || {};
    item('p', `${manifest.mvg_id} · ${coverage.mode || 'unknown'} · ${manifest.evidence.status} · blocking tasks: ${(coverage.blocking_task_ids || []).join(', ') || 'none'}`, flows);
    item('p', `Runtime evidence: ${manifest.evidence.reason || manifest.evidence.run_id || 'none'} · manifest SHA-256: ${manifest.sha256}`, flows);
    if (coverage.blocking_task_ids?.length) item('p', 'The manifest declares blocking tasks. Full playable scope is not cleared by a partial run.', flows);
    for (const flow of manifest.flows) {
      const row = document.createElement('details');
      item('summary', `${flow.id} · tasks ${flow.task_ids.join(', ') || 'unmapped'} · ${flow.version_ids.length ? flow.version_ids.length + ' traced version(s)' : 'no traced version'}`, row);
      item('p', flow.outcome, row);
      item('p', 'Related versions: ' + (flow.version_ids.join(' · ') || 'not mapped; cumulative flow remains in scope'), row);
      item('p', 'Tests: ' + (flow.test_ids.join(', ') || 'none declared'), row);
      for (const handoff of flow.handoffs || []) {
        const block = document.createElement('p'); item('strong', `Task ${handoff.producer_task} → ${handoff.consumer_task} · owner ${handoff.owner_task} · `, block);
        item('span', `${handoff.behavior || ''} · contract: `, block);
        if (handoff.contract_ref) source(handoff.contract_ref, block); else item('span', 'unmapped', block);
        item('span', ` · tests: ${(handoff.test_ids || []).join(', ') || 'none'}`, block); row.append(block);
      }
      flows.append(row);
    }
    list('Declared tests and evidence levels', manifest.tests.map(test => `${test.id} · ${test.evidence_level || test.state || 'unclassified'} · ${test.path || ''}`), flows);
    item('p', 'Human playcheck: verify the selected flow in Godot, the handoff behavior, save/continue or return path, and any dynamic routes. Static reachability and automated success do not establish subjective playability.', flows);
  }
  async function load() {
    get('mvg-status').textContent = 'Loading version evidence...';
    const response = await fetch('/api/knowledge/mvg-overview'); const payload = await response.json();
    if (!response.ok) throw new Error(payload.reason || 'Unable to load MVG evidence');
    overview = payload;
    for (const [id, rows, value, label] of [['mvg-version-select', payload.versions, 'id', 'gdd_path'], ['mvg-manifest-select', payload.manifests, 'path', 'mvg_id']]) {
      const selector = get(id); const previous = selector.value; selector.replaceChildren();
      for (const row of rows) { const option = item('option', row[label] + (value === 'id' ? ' @' + row.gdd_sha256.slice(0, 12) : ''), selector); option.value = row[value]; }
      if (rows.some(row => row[value] === previous)) selector.value = previous;
    }
    get('mvg-status').textContent = `${payload.versions.length} registered GDD candidates · ${payload.versions.filter(row => row.mapping === 'traced').length} traced · ${payload.manifests.length} cumulative MVG manifests · ${payload.topology_fresh ? 'fresh topology' : 'topology unavailable or stale'}. Historical GDD revisions require published provenance.`;
    render();
  }
  get('mvg-version-select').addEventListener('change', render);
  get('mvg-manifest-select').addEventListener('change', render);
  get('mvg-refresh').addEventListener('click', () => load().catch(error => { get('mvg-status').textContent = error.message; }));
  load().catch(error => { get('mvg-status').textContent = error.message; });
})();

File: scripts/python/_project_health_mvg_versions.py
"""Read-only, provenance-limited GDD version and cumulative MVG projection."""
from __future__ import annotations

import hashlib
import json
from pathlib import Path


def build_overview(snapshot, details: list[dict], topology: dict, graph: dict) -> dict:
    paths = set(snapshot.paths)
    registered = set()
    source_set = 'docs/workflows/chapter3-source-set.json'
    if source_set in paths:
        registered.update(json.loads(snapshot.read_text(source_set)).get('active_sources', []))
    # A configured GDD may precede the Chapter 3 source-set registry.
    if source_set not in paths:
        registered.update(path for path in paths if path.startswith('docs/gdd/') and path.endswith('.md'))
        registered.update(path for path in ('_bmad-output/gdd.md',) if path in paths)
    gdds = sorted(path for path in registered if path in paths and
                  (path.startswith('docs/gdd/') or path == '_bmad-output/gdd.md'))
    trustworthy = topology.get('available') is True and topology.get('fresh') is True and topology.get('identity', {}).get('revision') == snapshot.commit
    blocks = {}
    if trustworthy:
        for block in topology.get('nodes', {}).get('source_blocks', []):
            path = block.get('source_path')
            if path in gdds and (block.get('source_sha256') or block.get('source_file_sha256')) in (snapshot.digest(path), 'sha256:' + snapshot.digest(path)):
                blocks[str(block.get('block_id'))] = path
    task_ids = {str(row['task']['id']): row for row in details if isinstance(row.get('task'), dict)}
    associated = {path: set() for path in gdds}
    if trustworthy:
        for task_id, trace in topology.get('task_trace', {}).items():
            if task_id in task_ids:
                for block_id in trace.get('source_blocks', []):
                    if str(block_id) in blocks:
                        associated[blocks[str(block_id)]].add(task_id)

    versions = []
    for path in gdds:
        sha = snapshot.digest(path)
        mapped = sorted(associated[path], key=lambda item: (not item.isdigit(), int(item) if item.isdigit() else item))
        refs = {'overlays': set(), 'contracts': set(), 'adrs': set()}
        scenes = {}
        for task_id in mapped:
            detail = task_ids[task_id]
            for view in detail.get('mappings', {}).values():
                for row in view:
                    for field, key in (('overlay_refs', 'overlays'), ('contractRefs', 'contracts'), ('adr_refs', 'adrs')):
                        refs[key].update(value for value in row.get(field, []) if isinstance(value, str))
            for attachment in detail.get('godot', {}).get('scenes', []):
                scene_path = attachment.get('scene')
                if scene_path not in graph.get('nodes', {}):
                    continue
                scene = graph['nodes'][scene_path]
                scenes[scene_path] = {
                    'path': scene_path, 'classification': scene.get('classification'),
                    'nodes': sorted({str(node.get('parent') or '.') + '/' + str(node.get('name') or '(unnamed)') for node in scene.get('nodes', [])}),
                    'resources': sorted({resource for node in scene.get('nodes', []) for resource in node.get('resources', []) if isinstance(resource, str)}),
                    'evidence': 'declared_task_with_verified_static_attachment',
                }
        versions.append({
            'id': path + '@' + sha[:16], 'gdd_path': path, 'gdd_sha256': sha,
            'mapping': 'traced' if mapped else 'unmapped',
            'task_ids': mapped, 'tasks': [{'id': id_, 'title': task_ids[id_]['task'].get('title'), 'status': task_ids[id_]['task'].get('status')} for id_ in mapped],
            'references': {key: sorted(values) for key, values in refs.items()},
            'scenes': [scenes[key] for key in sorted(scenes)],
        })
    manifests = []
    for path in sorted(p for p in paths if p.startswith('docs/testing/mvg/') and p.endswith('.json')):
        data = snapshot.read_bytes(path)
        manifest = json.loads(data.decode('utf-8-sig'))
        if manifest.get('schema_version') != 'newrouge.mvg-integration.v1':
            continue
        flows = []
        for flow in manifest.get('flows', []):
            ids = {str(value) for value in flow.get('task_ids', [])}
            flows.append({'id': flow.get('id'), 'outcome': flow.get('outcome'),
                          'task_ids': sorted(ids), 'test_ids': flow.get('test_ids', []),
                          'handoffs': flow.get('handoffs', []),
                          'version_ids': [version['id'] for version in versions if ids.intersection(version['task_ids'])]})
        manifests.append({'path': path, 'sha256': 'sha256:' + hashlib.sha256(data).hexdigest(),
                          'mvg_id': manifest.get('mvg_id'), 'coverage': manifest.get('coverage', {}),
                          'flows': flows, 'tests': manifest.get('tests', [])})
    return {'schema_version': 'newrouge.knowledge-mvg-overview.v1', 'revision': snapshot.commit,
            'topology_fresh': trustworthy, 'versions': versions, 'manifests': manifests,
            'limitations': 'GDD source-set registration and static scene reachability do not prove Chapter 3 consumption or playability. Historical GDD revisions require separately published provenance.'}


def with_runtime_evidence(overview: dict, root: Path) -> dict:
    """Attach the latest matching run, never promoting a stale or workspace result."""
    result = {**overview, 'manifests': []}
    directory = root / 'logs/ci/mvg-acceptance'
    runs = sorted(directory.glob('*/summary.json'), key=lambda path: path.parent.name, reverse=True)[:200] if directory.exists() else []
    for manifest in overview.get('manifests', []):
        copy = dict(manifest)
        copy['evidence'] = {'status': 'not_verified', 'reason': 'No run matched this main revision and manifest content.'}
        for path in runs:
            try:
                summary = json.loads(path.read_text(encoding='utf-8'))
            except (OSError, ValueError):
                continue
            if (summary.get('manifest') != manifest['path'] or summary.get('manifest_sha256') != manifest['sha256']
                    or summary.get('mode') != 'run' or summary.get('source_revision') != overview.get('revision')
                    or summary.get('base_commit') != overview.get('revision') or summary.get('workspace_dirty') is not False):
                continue
            copy['evidence'] = {'status': 'passed' if summary.get('status') == 'passed' and summary.get('runtime_verified') is True else 'failed',
                                'run_id': summary.get('run_id'), 'reason': summary.get('reason', ''),
                                'runtime_verified': summary.get('runtime_verified') is True}
            break
        result['manifests'].append(copy)
    return result
